### PR TITLE
fix #278, and related

### LINKED ISSUE #278: rename latest version to min version
...

### PR DESCRIPTION
**Pull request by silentwings**

### CODE REVIEW

File: protocol/Protocol.py [hunk 1]
```diff
@@ -2667,15 +2676,24 @@ def in_ADMINBROADCAST(self, client, msg):
...
 
 		@required.str version: The new version to apply.
 		'''
-		self._root.latestspringversion = version
-		self.out_SERVERMSG(client, 'Latest spring version is now set to: %s' % version)
-
+		self._root.min_spring_version = version
+		self.in_BROADCAST(client, 'New engine version: Spring %s' % version)
+		
+		legacyBattleIds = []
```
Comment (abma) [inline protocol/Protocol.py]: why this? self.broadcast_RemoveBattle(battle) can be called directly?!

File: protocol/Protocol.py [hunk 2]
```diff
@@ -3134,7 +3152,7 @@ def getBanDB(self):
 	assert(not p._validChannelSyntax("#abcde")[0])
 	assert(not p._validChannelSyntax("ab cde")[0])
 
-	minver = "104.0"
+	p._root.min_spring_version = "104.0"
```
Comment (abma) [inline protocol/Protocol.py]: why this change? in this case "104.0" is specific for the tests.

File: protocol/Protocol.py [hunk 3]
```diff
@@ -649,7 +652,13 @@ def clientFromSession(self, session_id):
 		logging.error("Couldn't get client from session_id: %d %s"% (session_id, traceback.format_exc()))
 		self.cleanup()
 		return None
-
+	
+	def hasBotflag(self, battle):
```
Comment (abma) [inline protocol/Protocol.py]: this function looks useless...

battle.host.bot would do the same

File: protocol/Protocol.py [hunk 4]
```diff
@@ -188,23 +188,16 @@ def versiontuple(version):
...
 
-def validEngineVersion(engine, version, minver):
-	if engine != "spring":
-		return False
-	if not version:
-		return False
-	return versiontuple(version) >= versiontuple(minver)
 
-	
 flag_map = {
```
Comment (abma) [inline protocol/Protocol.py]: why the whitespace changes? this breaks layout

File: DataHandler.py [hunk 5]
```diff
@@ -304,6 +308,18 @@ def parseFiles(self):
 		ins.close()
 		self.loadCertificates()
 
+	def set_latest_spring_version(self):
```
Comment (abma) [inline DataHandler.py]: IMHO this should be removed: the benefit is very low and it could cause weird behaviour.

File: protocol/Protocol.py (hunk 1 above)
Comment (silentwings) [reply protocol/Protocol.py]: You can't remove elements from a dictionary while iterating through it.

File: protocol/Protocol.py (hunk 2 above)
Comment (silentwings) [reply protocol/Protocol.py]: Because I thought it was cleaner not to have to pass the current min version in to _validEngineVersion, and make _validEngineVersion a method of the procotol class.

The root here is a dummy, so afaics there is nothing odd about this as unit test code.

File: protocol/Protocol.py (hunk 4 above)
Comment (silentwings) [reply protocol/Protocol.py]: accident, will fix

File: DataHandler.py (hunk 5 above)
Comment (silentwings) [reply DataHandler.py]: I thought you might appreciate not having to call SETMINSPRINGVERSION every time the server was restarted. Since you don't want it (and I have no use for it), will remove.

This code is essentially identical to https://github.com/spring/uberserver/blob/master/DataHandler.py#L427. I don't know what is meant by "wierd behaviour".

Comment (abma) [reply DataHandler.py]: > I don't know what is meant by "wierd behaviour".

undefined / non-deterministic behaviour. using this param could cause the server not to start up if the webserver is not reachable. also the min version set very likely is not the version which should be set when using this param / function.

Comment (silentwings) [reply DataHandler.py]: If this could cause the server not to start, the same issue presumably exists with the existing detectIp function, which contacts the same webserver. (I don't think this happens though - when I tested it with my internet disconnected, it just waited for the timeout period set and then started as normal.)

File: protocol/Protocol.py (hunk 2 above)
Comment (silentwings) [reply protocol/Protocol.py]: Re-open if you disagree!

Comment (gajop) [reply protocol/Protocol.py]: As I understand it, 104.x doesn't run on Mac, while 103.x does. If I'm not wrong, I can understand why people might complain with enforcing 104.x

Comment (abma) [reply protocol/Protocol.py]: we need to move forward... apple has to decided to deprecate opengl. very likely newer version of osx will break spring 103.x, too.

Comment (silentwings) [reply protocol/Protocol.py]: This piece of code is a unit test - not a hardcoded default. (I don't see any future worth the trouble for macs, though.)

Comment (gajop) [reply protocol/Protocol.py]: I see - my bad for looking at just code snippets.

Comment (abma) [reply protocol/Protocol.py]: @gajop
deprecating usually means that it will be removed after some time.

also 104.0 is currently planned as min version, so your concerns are valid.

**Comment by abma** [comment]
only one concern left:

not sure if the autoclose of battles is good: IMHO it should send at least a message to the battle before closing so users / autohosts have a chance to understand what happened.

**Comment by silentwings** [comment]
I think the battles need to be closed, otherwise autohosts could stay open for weeks or even months.

Battles will get the same server message as is sent to everyone, but I'll add a specific message for each battle too.

File: protocol/Protocol.py [hunk 6]
```diff
@@ -2689,6 +2689,8 @@ def in_SETMINSPRINGVERSION(self, client, version):
 		for battleId, battle in self._root.battles.items():
 			if self.hasBotflag(battle) and not self._validEngineVersion(battle.engine, battle.version):
 				legacyBattleIds.append(battleId)
+				host = self.clientFromSession(battle.host)
+				self.broadcast_SendBattle(battle, 'SAIDBATTLEEX %s This battle will close, please join a battle with the new Spring version!' % host.username)
```
Comment (abma) [inline protocol/Protocol.py]: shouldn't this be sent by ChanServ?

**Comment by abma** [comment]
IMHO the message could be better, maybe sth. like this?

"Battle closed because the minimum allowed engine version was set to %s %s which is larger then the used version."
